qiskit/algorithms/linear_solvers/matrices/linear_system_matrix.py

```diff
@@ -0,0 +1,137 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""An abstract class for linear systems solvers in Qiskit's aqua module."""
+
+from abc import ABC, abstractmethod
+
+from qiskit import QuantumCircuit
+from qiskit.circuit.library import BlueprintCircuit
+
+
+class LinearSystemMatrix(BlueprintCircuit, ABC):
+    """Base class for linear system matrices."""
+
+    def __init__(self, num_state_qubits: int, tolerance: float, evo_time: float,
+                 name: str = 'ls_matrix') -> None:
+        """
+        Args:
+            num_state_qubits: the number of qubits where the unitary acts.
+            tolerance: the accuracy desired for the approximation
+            evo_time: the time of the Hamiltonian simulation
+            name: The name of the object.
+        """
+        super().__init__(name=name)
+
+        # define internal parameters
+        self._num_state_qubits = None
+        self._tolerance = None
+        self._evo_time = None  # makes sure the eigenvalues are contained in [0,1)
+
+        # store parameters
+        self.num_state_qubits = num_state_qubits
+        self.tolerance = tolerance
+        self.evo_time = evo_time
+
+    @property
+    def num_state_qubits(self) -> int:
+        r"""The number of state qubits representing the state :math:`|x\rangle`.
+
+        Returns:
+            The number of state qubits.
+        """
+        return self._num_state_qubits
+
+    @num_state_qubits.setter
+    def num_state_qubits(self, num_state_qubits: int) -> None:
+        """Set the number of state qubits.
+
+        Note that this may change the underlying quantum register, if the number of state qubits
+        changes.
+
+        Args:
+            num_state_qubits: The new number of qubits.
+        """
+        if num_state_qubits != self._num_state_qubits:
+            self._invalidate()
+            self._num_state_qubits = num_state_qubits
+            self._reset_registers(num_state_qubits)
+
+    @property
+    def tolerance(self) -> float:
+        """Return the error tolerance"""
+        return self._tolerance
+
+    @tolerance.setter
+    def tolerance(self, tolerance: float) -> None:
+        """Set the error tolerance
+        Args:
+            tolerance: The new error tolerance.
+        """
+        self._tolerance = tolerance
+
+    @property
+    def evo_time(self) -> float:
+        """Return the time of the evolution."""
+        return self._evo_time
+
+    @evo_time.setter
+    def evo_time(self, evo_time: float) -> None:
+        """Set the time of the evolution.
+
+        Args:
+            evo_time: The new time of the evolution.
+        """
+        self._evo_time = evo_time
+
+    @property
+    def eigs_bounds(self) -> [float, float]:
+        """Return lower and upper bounds on the eigenvalues of the matrix."""
+        return None
```

Comment:
I imagine the intent of returning None is since this a base class it needs to be overridden. As such it would be much better to decorate this as an abstract method. Otherwise a derived class can fail to override this and if this were called it would return None, which is not a valid return value since the return typehint is not Optional.  And for the return type it should be Tuple[float, float] ?

Same for condition_bounds. And as that is a bit more readable would eigenvalue_bounds be better as a name. Though I agree eig and eigs are used in mupy.scipy so the abbreviation is common

Reply:
Changed to Tuple, however for some problems it won't be possible to override the methods. For the matrices considered so far there are analytic formulas to calculate their eigenvalues, but in general this is not possible. The HHL algorithm checks whether the matrix class given as an input contains these methods, and if it does uses the information to make the computation a bit more efficient.

Reply:
I would argue you can always override the method. Otherwise the user of such a class you describe, if they do call this on such an instance, will get back None from this method which is likely to lead to some failure that may be hard to track. Better in such a class, that cannot support this method, to override and document that is not something that can be implemented and also raise a NotImplementedError with an equivalent error message in it so that if the user does end up calling it they will get an immediate failure with a meaningful message of what they did wrong. I hesitate to suggest different base classes, one without these methods in it and one with, that inherited from the former, so an algorithm could extend either; but that would be doable too.

Reply:
Made an abstract method